### PR TITLE
Documentation for PoC on opaque integer resources.

### PR DESCRIPTION
**Pull request by balajismaniam**
1. Added description on the design and implementation. 
2. Added simple usage example.

### CODE REVIEW

File: docs/proposals/opaque-integer-resources.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+curl --header "Content-Type:application/json-patch+json" \
+        --request PATCH \
+        --data '[{"op": "add", "path": "/status/capacity/opaque-int.alpha.kubernetes.io~1<resource-name>", "value": "<integer-value>"}]' \
+        http://<apiserver-host>:<apiserver-port>/api/v1/nodes/<node-name>/status
+```
+
+where,
+
+- `<resource-name>` is the name of the opaque integer resource being advertised. 
+- `<integer-value>` is the number of the opaque integer resource available on the node. This value must be a integer.
```
Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: s/a/an

File: docs/proposals/opaque-integer-resources.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+where,
+
+- `<resource-name>` is the name of the opaque integer resource being advertised. 
+- `<integer-value>` is the number of the opaque integer resource available on the node. This value must be a integer. 
+- `<apiserver-host>` is the hostname or the IP of the kube-apiserver. 
+- `<apiserver-port>` is the port on which the kube-apiserver is listening. 
+- `<node-name>` is the name of node for which we want to update the capacity.
+
+With apporpriate changes, the above script will update the capcity of the node. In the next sync with the api server, the kubelet corresponding to the node will update its own `capacity` and `allocatable`. Subsequently, the scheduler on its next sync with the api server updates the value of allocatable resource in its own cache for that particular.
```
Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: s/apporpriate/appropriate

Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: s/capcity/capacity

Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: for that particular... missing end of sentence looks like.

File: docs/proposals/opaque-integer-resources.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+- `<apiserver-host>` is the hostname or the IP of the kube-apiserver. 
+- `<apiserver-port>` is the port on which the kube-apiserver is listening. 
+- `<node-name>` is the name of node for which we want to update the capacity.
+
+With apporpriate changes, the above script will update the capcity of the node. In the next sync with the api server, the kubelet corresponding to the node will update its own `capacity` and `allocatable`. Subsequently, the scheduler on its next sync with the api server updates the value of allocatable resource in its own cache for that particular. 
+
+
+### Consuming Opaque Integer Resources
+
+After the advertisement of the opaque integer resource as explained above, it is ready to be consumed like a regular resource (e.g., CPU or memory). The following spec shows how to configure a pod to consume a opaque integer resource called `resource1`. Assuming that there are enough resource available to be allocated, the scheduler will schedule this pod to the appropriate node in the Kubernetes cluster.
```
Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: s/a opaque/an opaque

Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: s/enough resource/enough resources

Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: s/schedule this pod/bind this pod
(?)

File: docs/proposals/opaque-integer-resources.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        http://<apiserver-host>:<apiserver-port>/api/v1/nodes/<node-name>/status
+```
+
+where,
+
+- `<resource-name>` is the name of the opaque integer resource being advertised. 
+- `<integer-value>` is the number of the opaque integer resource available on the node. This value must be a integer. 
+- `<apiserver-host>` is the hostname or the IP of the kube-apiserver. 
+- `<apiserver-port>` is the port on which the kube-apiserver is listening. 
+- `<node-name>` is the name of node for which we want to update the capacity.
```
Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: s/name of node/name of the node

File: docs/proposals/opaque-integer-resources.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
+# Accounting and Consuming Opaque Integer Resources
+
```
Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md:27]: Could we add a ToC?

File: docs/proposals/opaque-integer-resources.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+## Terms
+
+### Opaque node-level resource
+A resource associated with a node that has a consumable capacity, which is handled uniformly by the scheduler and Kubelet. (If there are specializations that hang off of the resource name, then it is not opaque!)
+
+## Requirements
+- Associate available opaque counted resources with existing nodes.
```
Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: Maybe a numbered list would be better here so people can refer to the requirements that way.

File: docs/proposals/opaque-integer-resources.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+Even before implementing isolation, there is already user value in accounting opaque resources in the cluster.
+
+* Exclusive access to discrete co-processors (GPGPUs, MICs, FPGAs, etc.)
+* Counting slots allowed to access a shared parallel file system.
+* Support for tracking additional cgroup quotas like `cpu_rt_runtime`.
+
+## Goals
+1. Associate opaque integer resources with nodes.
+2. Request some portion of available capacity in the Pod spec during pod creation.
```
Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: normalize capitalization of "pod" (lowercase elsewhere in this doc).

File: docs/proposals/opaque-integer-resources.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+3. Enable users to define and schedule on opaque resources in a vanilla Kubernetes cluster.
+
+## Non-goals
+1. Solve how opaque resource allocations are to be enforced/isolated.
+2. Solve how opaque resources are discovered, beyond existing APIs.
+3. Solve how cluster-level opaque resources should be advertised.
```
Comment (ConnorDoyle) [inline docs/proposals/opaque-integer-resources.md]: Wondering if this is still a non-goal, given the design discussion below. Seems this is covered.

**Comment by ConnorDoyle** [comment]
Let's also pull in this section (added after you did the first transcription) https://docs.google.com/document/d/1AeSBHqLVTLanHBxUBJ3J3CbXp50a-d7cmY5bAh2Bahc/edit#bookmark=id.gz5xm7vh9vq0

**Comment by ConnorDoyle** [comment]
Thanks, just commented on minor changes. Otherwise LGTM

File: docs/proposals/opaque-integer-resources.md (hunk 8 above)
Comment (balajismaniam) [reply docs/proposals/opaque-integer-resources.md]: Yes. The advertising part is covered below. I will remove it from here. But we should make it clear we want to deal only with node-level resources for now. I will make appropriate changes to the document.

**Comment by balajismaniam** [comment]
@ConnorDoyle: I've updated the doc. PTAL.

**Comment by ConnorDoyle** [comment]
Looks great for starters. We can iterate it down the line. Merging.

**Pull request merged**